### PR TITLE
Move the ACLs out of the group model and into their own file

### PR DESCRIPTION
**Pull request by jon-betts**
This is not a desired end-state. The idea is to collect all ACL behavior into one place.

This needs quite a few things to be in the right shape before this can happen:

 * https://github.com/hypothesis/h/pull/6866
 * https://github.com/hypothesis/h/pull/6867
 * https://github.com/hypothesis/h/pull/6875
 * https://github.com/hypothesis/h/pull/6870

Turns out there were a few missed parts where the group was still used. These disturbingly passed just fine with the change (which they really shouldn't have). It turns out our tests often setup the return value for `has_permission` but make no assertions about how it's called.

This isn't good, but it's actually to our advantage for the moment, as these will likely get swapped out for something else at some point.

### CODE REVIEW

File: tests/h/models/group_test.py [hunk 1]
```diff
@@ -238,259 +238,6 @@ def test_non_public_group():
     assert not group.is_public
 
 
-class TestGroupACL:
-    def test_auth_client_with_matching_authority_may_read_members(
-        self, group, authz_policy
-    ):
-        group.authority = "weewhack.com"
```
Comment (jon-betts) [inline tests/h/models/group_test.py:245]: All of this weird "weewack.com" stuff is gone in favour of using `f"client_authority:{group.authority}"`

File: tests/h/models/group_test.py [hunk 2]
```diff
@@ -238,259 +238,6 @@ def test_non_public_group():
...
-    def authz_policy(self):
-        return ACLAuthorizationPolicy()
-
-    @pytest.fixture
-    def group(self):
-        creator = models.User(username="luke", authority="example.com")
-        group = models.Group(
-            name="test-group", authority="example.com", creator=creator
-        )
-        group.pubid = "test-group"
```
Comment (jon-betts) [inline tests/h/models/group_test.py:490]: All of these magic values have been replaced with interpolation instead, which allows this to be replaced with some simple factories.

File: tests/h/security/test_acl.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        def permitted_principals_for(permission):
+            class ACLCarrier:
+                def __acl__(self):
+                    return ACL.for_group(group)
+
+            return ACLAuthorizationPolicy().principals_allowed_by_permission(
+                ACLCarrier(), permission
+            )
+
+        return permitted_principals_for
```
Comment (jon-betts) [inline tests/h/security/test_acl.py:159]: Eh some duplication here, but it's not the end of the world. With luck we're getting rid of this whole system soon.

File: h/security/acl.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        # group
+        yield Allow, role.Staff, Permission.Group.ADMIN
+        yield Allow, role.Admin, Permission.Group.ADMIN
+
+        if group.creator:
+            yield Allow, group.creator.userid, Permission.Group.ADMIN
+            yield Allow, group.creator.userid, Permission.Group.MODERATE
+            yield Allow, group.creator.userid, Permission.Group.UPSERT
+
+        yield security.DENY_ALL
```
Comment (jon-betts) [inline h/security/acl.py:55]: Notice the slightly different grouping here. I've tried to group things by multiple things at once. 

So you can see all of the ADMIN privileges are next to each other, but all of the creator and client authority ones are grouped as well.

It's not perfect, but I've tried to keep as many concepts local as possible.

File: h/traversal/group.py [hunk 5]
```diff
@@ -53,4 +54,5 @@ def __acl__(self):
 
         # If there is a group associated with the context, the "upsert" and
         # all other permissions are managed by the model
-        return self.group.__acl__()
+
+        return ACL.for_group(self.group)
```
Comment (jon-betts) [inline h/traversal/group.py:58]: The stuff here will get merged into the ACL class soon.

File: h/storage.py [hunk 6]
```diff
@@ -126,10 +121,10 @@ def create_annotation(request, data, group_service):
     # further checks.
     group = group_service.find(data["groupid"])
     if group is None or not request.has_permission(
-        Permission.Group.WRITE, context=group
+        Permission.Group.WRITE, context=GroupContext(group)
     ):
         raise schemas.ValidationError(
-            "group: " + _("You may not create annotations " "in the specified group!")
+            "group: " + _("You may not create annotations in the specified group!")
```
Comment (jon-betts) [inline h/storage.py:127]: This super sucks for grepping for things

File: h/security/acl.py (hunk 4 above)
Comment (marcospri) [reply h/security/acl.py:55]: Yup, also the generator seems somehow easier to read, I guess `terms.append(` adds more noise than just `yield` :+1:

**Review by marcospri — APPROVED** [review]
I haven't done any manual testing just compared `h/models/group.py` to `h/security/acl.py` and that looks good.

**Comment by jon-betts** [comment]
The fond hope now is the there is actually enough coverage to notice mistakes in this stuff. I'll give it some light manual testing in QA and Live still though.

**Pull request merged**